keps/sig-node/4010-head-request-probes/README.md

```diff
@@ -0,0 +1,265 @@
+
+#### Unit tests
+
+- Add unit tests to verify the correct handling of HEAD requests in the Kubernetes API server and Kubelet components.
+- Test the behavior of the probes when HEAD requests are used.
+
+#### Integration tests
+
+- Perform integration tests to validate the interoperability of the updated components with other Kubernetes functionality.
+- Verify the correct behavior of HEAD-based probes in various scenarios, including failures and recoveries.
+
+#### e2e tests
+
+- Create end-to-end tests that cover the entire lifecycle of pods with HEAD-based probes.
+- Validate the behavior of probes in real-world scenarios, ensuring they function as expected.
+
+### Graduation Criteria
+
+The support for HEAD requests in readiness and liveness probes will undergo the following graduation stages:
+
+1. Alpha: v1.28, enabled via a feature flag. (Q: are we doing this with GA APIs?)
+2. Beta: v1.29
+3. GA: v1.30
+
+### Upgrade / Downgrade Strategy
+
+To upgrade to a version of Kubernetes that supports HEAD requests in probes, users will need to update their API server, Kubelet, and client libraries. Compatibility checks should be performed during the upgrade
+
+### Version Skew Strategy
+
+To manage version skew during the introduction of support for HTTP HEAD requests in readiness and liveness probes, the following strategy will be implemented:
+
+- Documentation: Clear documentation will be provided that outlines the supported Kubernetes versions for the new feature. This documentation will include information on compatibility between different Kubernetes components.
+
+- Compatibility Checks: During the upgrade process, compatibility checks will be performed to ensure that the Kubernetes cluster components are running compatible versions. These checks will help identify any version skew issues and provide guidance on resolving them before enabling the new feature.
+
+## Production Readiness Review Questionnaire
+
+Before enabling support for HTTP HEAD requests in readiness and liveness probes in a production environment, the following aspects will be assessed through a readiness review questionnaire:
+
+### Feature Enablement and Rollback
+
+- What flags or configuration settings are required to enable the new feature?
+
+A feature flag `HeadRequestProbes` will be introducded.
+
+- Can the feature be disabled or rolled back if issues arise?
+
+Yes.
+
+- Are there any specific considerations for enabling or disabling the feature on a live cluster?
+
+No.
+
+### Rollout, Upgrade, and Rollback Planning
+
+- What is the recommended rollout strategy for enabling the new feature?
+
+Update to the version which has the new feature implemented and enable the feature gate.
+
+- Are there any specific considerations or steps to follow during the rollout process?
+
+No.
+
+- Are there any backward compatibility concerns to address during the rollout?
+
+No.
```

Comment:
Is that right?

(How would someone roll back if they have defined HTTP `HEAD` probes on existing Pods or Pod templates?)

Reply:
This is a very real issue - it's not at all uncommon to have nodes on a back-rev release vs. apiserver.

One option is "too bad, so sad, upgrade your nodes" - we can't even easily notify you (I suppose we could throw evens or something we we found a probe, but it had no mechanism defined, but I don't think we do that today).

Another is to bend the API so that "use HEAD" is an extension of the existing HTTP mechanism (sadly named "httpGet"), so old kubelets could fall back on GET.

Reply:
I think they will see the error: 

https://github.com/kubernetes/kubernetes/blob/4706516100649bf5a92c780488ea53924239ad17/pkg/kubelet/prober/prober.go#L181-L182

So discoverability will be OK

Reply:
We might at least want to document the rollback if the KEP reached beta.